pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -1228,57 +1260,145 @@ func updatePipelineRunStatusFromTaskRuns(logger *zap.SugaredLogger, pr *v1beta1.
+		cr, inChildRef := childRefByPipelineTask[pipelineTaskName]
+		if inChildRef {
+			taskRunName = cr.Name
+		} else if taskRunName, ok = taskRunByPipelineTask[pipelineTaskName]; !ok {
 			// The pipelineTask associated to the conditions was not found in the pipelinerun
 			// status. This means that the conditions were orphaned, and never added to the
 			// status. In this case we need to generate a new TaskRun name, that will be used
 			// to run the TaskRun if the conditions are passed.
-			taskRunName = resources.GetTaskRunName(pr.Status.TaskRuns, pipelineTaskName, pr.Name)
+			taskRunName = resources.GetTaskRunName(pr.Status.TaskRuns, pr.Status.ChildReferences, pipelineTaskName, pr.Name)
+		}
+
+		if _, ok := childRefByPipelineTask[pipelineTaskName]; !ok && minimalEmbedded {
+			childRefByPipelineTask[pipelineTaskName] = &v1beta1.ChildStatusReference{
+				TypeMeta: runtime.TypeMeta{
+					APIVersion: v1beta1.SchemeGroupVersion.String(),
+					Kind:       "TaskRun",
+				},
+				Name:             taskRunName,
+				PipelineTaskName: pipelineTaskName,
+			}
+		}
+
+		if _, ok := taskRunByPipelineTask[pipelineTaskName]; !ok && fullEmbedded {
 			pr.Status.TaskRuns[taskRunName] = &v1beta1.PipelineRunTaskRunStatus{
 				PipelineTaskName: pipelineTaskName,
 				Status:           nil,
 				ConditionChecks:  nil,
 			}
 		}
-		// Build the map of condition checks for the taskrun
-		// If there were no other condition, initialise the map
-		conditionChecks := pr.Status.TaskRuns[taskRunName].ConditionChecks
-		if conditionChecks == nil {
-			conditionChecks = make(map[string]*v1beta1.PipelineRunConditionCheckStatus)
+
+		updateConditionChecksForTaskRun(logger, pr, childRefByPipelineTask, actualConditionTaskRuns, pipelineTaskName, taskRunName, fullEmbedded, minimalEmbedded)
+	}
+	if minimalEmbedded {
+		var newChildRefs []v1beta1.ChildStatusReference
+		for k := range childRefByPipelineTask {
+			newChildRefs = append(newChildRefs, *childRefByPipelineTask[k])
 		}
-		for i, foundTaskRun := range actualConditionTaskRuns {
-			lbls := foundTaskRun.GetLabels()
-			if _, ok := conditionChecks[foundTaskRun.Name]; !ok {
-				// The condition check was not found, so we need to add it
-				// We only add the condition name, the status can now be gathered by the
-				// normal reconcile process
-				if conditionName, ok := lbls[pipeline.ConditionNameKey]; ok {
-					conditionChecks[foundTaskRun.Name] = &v1beta1.PipelineRunConditionCheckStatus{
-						ConditionName: fmt.Sprintf("%s-%s", conditionName, strconv.Itoa(i)),
-					}
-				} else {
-					// The condition name label is missing, so we cannot recover this
-					logger.Warnf("found an orphaned condition taskrun %#v with missing %s label",
-						foundTaskRun, pipeline.ConditionNameKey)
+		pr.Status.ChildReferences = newChildRefs
+	}
+}
+
+func updateConditionChecksForTaskRun(logger *zap.SugaredLogger, pr *v1beta1.PipelineRun, childRefByPipelineTask map[string]*v1beta1.ChildStatusReference,
```

Comment:
I'll add a comment, but again, when I tried rewriting this to tease apart `TaskRuns` and `ChildReferences` usages completely, it was faaaaar messier than what's there now.

Reply:
...which is not to say it's impossible to make it less messy, of course, just that I can't seem to pull that off. I can try again later, and almost certainly will, but I think that if the messiness of these functions ends up as the last thing holding this PR up, this may be a case where we need to plug our nose and accept the smell.

Reply:
and fwiw, the messiness feels to me to largely be down to the fact that `pr.Status.TaskRuns` and `pr.Status.Runs` are maps, while `pr.Status.ChildReferences` is an array. So while updating `pr.Status.TaskRuns` and `pr.Status.Runs` is just a matter of "oh, this key isn't there yet? Let's add it! Oh, it is? Then update its value!" without having to worry about anything else already in the map, updating `pr.Status.ChildReferences` means we need to make sure that everything already in the array is still in the array when we're done, and we have to make changes to the elements in the array at different points in the process. That's why I added the `childRefByPipelineTask` temporary map - so we can keep all the existing references, add new ones, and make changes to references without needing to do "ok, iterate over `pr.Status.ChildReferences` until we find one with the right `PipelineTaskName`, now make changes to it, etc".

That last bit would admittedly be simpler if `pr.Status.ChildReferences` was a `[]*ChildStatusReference` rather than a `[]ChildStatusReference` as it is now, but it'd still be a mess.

It's definitely better from a Kubernetes API perspective to have an array instead of a map, but the downside is that updating individual elements in an array is always going to be uglier than updating individual elements in a map.